profiling/perfetto-connector/libperfetto-connector.cpp

```diff
@@ -0,0 +1,182 @@
+
+template <typename Tp>
+auto get_env(std::string_view _env_name, Tp _default) {
+  const char* _env_val = std::getenv(_env_name.data());
+
+  if (_env_val) {
+    std::stringstream _ss{};
+    _ss << std::skipws << _env_val;
+    if constexpr (std::is_same<std::decay_t<Tp>, const char*>::value) {
+      std::stringstream _ss{};
+      _ss << _env_val;
+      return _ss.str();
+    } else {
+      Tp _val{};
+      std::istringstream{_env_val} >> _val;
+      return _val;
+    }
+  }
+
+  if constexpr (std::is_same<std::decay_t<Tp>, const char*>::value) {
+    return std::string{_default};
+  } else {
+    return _default;
+  }
+}
+}  // namespace
+
+extern "C" {
+
+void kokkosp_init_library(const int loadSeq, const uint64_t interfaceVer,
+                          const uint32_t, void*) {
+  printf("-----------------------------------------------------------\n");
+  printf("KokkosP: Perfetto Connector (sequence is %d, version: %lu)\n",
+         loadSeq, interfaceVer);
+  printf("-----------------------------------------------------------\n");
+
+  perfetto::TracingInitArgs args{};
+  perfetto::TraceConfig cfg{};
+  perfetto::protos::gen::TrackEventConfig track_event_cfg{};
+
+  // default 10 MB shared memory and 1 GB buffer
+  auto shmem_size_hint = get_env("KOKKOSP_PERFETTO_SHMEM_SIZE_HINT_KB", 10000);
+  auto buffer_size = get_env("KOKKOSP_PERFETTO_BUFFER_SIZE_KB", 1000 * 1000);
+  auto _fill_policy_name = get_env("KOKKOSP_PERFETTO_FILL_POLICY", "discard");
+
+  if (std::getenv("KOKKOSP_PERFETTO_INPROCESS_BACKEND") &&
+      std::stoi(std::getenv("KOKKOSP_PERFETTO_INPROCESS_BACKEND")) != 0)
+    args.backends |= perfetto::kInProcessBackend;
+  if (std::getenv("KOKKOSP_PERFETTO_SYSTEM_BACKEND") &&
+      std::stoi(std::getenv("KOKKOSP_PERFETTO_SYSTEM_BACKEND")) != 0)
+    args.backends |= perfetto::kInProcessBackend;
+
+  if (args.backends == 0) args.backends |= perfetto::kInProcessBackend;
+
+  auto _fill_policy =
+      _fill_policy_name == "discard"
+          ? perfetto::protos::gen::TraceConfig_BufferConfig_FillPolicy_DISCARD
+          : perfetto::protos::gen::
+                TraceConfig_BufferConfig_FillPolicy_RING_BUFFER;
+
+  auto* buffer_config = cfg.add_buffers();
+  buffer_config->set_size_kb(buffer_size);
+  buffer_config->set_fill_policy(_fill_policy);
+
+  auto* ds_cfg = cfg.add_data_sources()->mutable_config();
+  ds_cfg->set_name("track_event");  // this MUST be track_event
+  ds_cfg->set_track_event_config_raw(track_event_cfg.SerializeAsString());
+
+  args.shmem_size_hint_kb = shmem_size_hint;
+
+  perfetto::Tracing::Initialize(args);
+  perfetto::TrackEvent::Register();
+
+  tracing_session = perfetto::Tracing::NewTrace();
+  tracing_session->Setup(cfg);
+  tracing_session->StartBlocking();
+}
+
+void kokkosp_finalize_library() {
+  printf("-----------------------------------------------------------\n");
+  printf("KokkosP: Finalizing Perfetto Connector... ");
+  fflush(stdout);
```

Comment:
Aww man I was so pleased with the environment parsing and now we are back to the good ole printf?

Reply:
🤣 I just tweaked this from David's. I used to never use it but, over time, I've come to appreciate it more. I'd replace it with std::format in a heartbeat.